### PR TITLE
Remove recursion from xlat_tables_v2 library

### PR DESCRIPTION
**Pull request by DavidPu**
This PR converts recursive translation table traverses to simple stack based version. It keeps exactly same translation table traversal order compared to recursive version.

fixes arm-software/tf-issues#664

### CODE REVIEW

**Comment by antonio-nino-diaz-arm** [comment]
First comments, I haven't checked the code but I will do it today.

There is a compilation error:
```
lib/xlat_tables_v2/xlat_tables_core.c: In function ‘xlat_tables_map_region’:
lib/xlat_tables_v2/xlat_tables_core.c:673:5: error: ‘subtables’ undeclared (first use in this function); did you mean ‘subtable’?
     subtables = (uintptr_t)(desc & TABLE_ADDR_MASK);
     ^~~~~~~~~
     subtable
```

Also, I've run my tests (I've just pushed them here https://review.trustedfirmware.org/#/q/topic:an/xlat-v2-tests+(status:open+OR+status:merged) ). They pass after applying your changes, so I'm reasonably sure that the changes are correct.

**Comment by antonio-nino-diaz-arm** [comment]
IP Review +1

**Comment by antonio-nino-diaz-arm** [comment]
Excluding the problem I've mentioned in https://github.com/ARM-software/arm-trusted-firmware/pull/1842 and the problems mentioned here it looks fine.

**Comment by DavidPu** [comment]
> First comments, I haven't checked the code but I will do it today.
> 
> There is a compilation error:
> 
> ```
> lib/xlat_tables_v2/xlat_tables_core.c: In function ‘xlat_tables_map_region’:
> lib/xlat_tables_v2/xlat_tables_core.c:673:5: error: ‘subtables’ undeclared (first use in this function); did you mean ‘subtable’?
>      subtables = (uintptr_t)(desc & TABLE_ADDR_MASK);
>      ^~~~~~~~~
>      subtable
> ```
> Also, I've run my tests (I've just pushed them here https://review.trustedfirmware.org/#/q/topic:an/xlat-v2-tests+(status:open+OR+status:merged) ). They pass after applying your changes, so I'm reasonably sure that the changes are correct.

Thanks for reviewing and testing! build break is fixed with updated PR, uint32_t data type is changed to 'unsigned int' to align with latest upstream code.

I also created a host side unittest application to compare(before and after apply these changes)  xlat_table dump and CRC values[1] for key xlat_ctx data structures. I will push it to separate repo.

[1] CRC is a little bit tricky for some data structure which stores pointers, I have to make sure the data structure is allocated at same address for different binaries.

**Comment by antonio-nino-diaz-arm** [comment]
Can you rebase this PR and fix the conflicts?

**Comment by DavidPu** [comment]
> Can you rebase this PR and fix the conflicts?
Done.

File: lib/xlat_tables_v2/xlat_tables_core.c [hunk 1]
```diff
@@ -537,105 +537,169 @@ static action_t xlat_tables_map_region_action(const mmap_region_t *mm,
 }
 
 /*
- * Recursive function that writes to the translation tables and maps the
+ * Function that writes to the translation tables and maps the
  * specified region. On success, it returns the VA of the last byte that was
- * successfully mapped. On error, it returns the VA of the next entry that
+ * succesfully mapped. On error, it returns the VA of the next entry that
```
Comment (antonio-nino-diaz-arm) [inline lib/xlat_tables_v2/xlat_tables_core.c]: Typo.

File: lib/xlat_tables_v2/xlat_tables_utils.c [hunk 2]
```diff
@@ -118,80 +118,137 @@ static void xlat_tables_print_internal(xlat_ctx_t *ctx, uintptr_t table_base_va,
...
+	size_t level_size = XLAT_BLOCK_SIZE(this_level);
 	unsigned int table_idx = 0U;
-	size_t level_size = XLAT_BLOCK_SIZE(level);
+	uintptr_t table_idx_va = table_base_va;
 
 	/*
 	 * Keep track of how many invalid descriptors are counted in a row.
 	 * Whenever multiple invalid descriptors are found, only the first one
 	 * is printed, and a line is added to inform about how many descriptors
+
```
Comment (antonio-nino-diaz-arm) [inline lib/xlat_tables_v2/xlat_tables_utils.c]: remove this empty line

Comment (DavidPu) [reply lib/xlat_tables_v2/xlat_tables_utils.c]: done

File: lib/xlat_tables_v2/xlat_tables_core.c (hunk 1 above)
Comment (DavidPu) [reply lib/xlat_tables_v2/xlat_tables_core.c]: done.

**Comment by antonio-nino-diaz-arm** [comment]
jenkins: test this please

**Comment by antonio-nino-diaz-arm** [comment]
This fails to build in AArch32 builds. For example, for this one:
```
AARCH32_SP=optee
ARCH=aarch32
ARM_ARCH_MAJOR=7
CROSS_COMPILE=arm-linux-gnueabihf-
PLAT=qemu
```
The output is:
```
lib/xlat_tables_v2/xlat_tables_core.c: In function 'xlat_tables_map_region':
lib/xlat_tables_v2/xlat_tables_core.c:720:40: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]
    this_base[table_idx] = TABLE_DESC | (uint64_t)subtable;
```

**Comment by DavidPu** [comment]
sorry, forgot to try AArch32 before. I just updated the PR to fix it.

**Comment by antonio-nino-diaz-arm** [comment]
jenkins: test this please

**Comment by soby-mathew** [comment]
jenkins: test this please

**Comment by antonio-nino-diaz-arm** [comment]
This fails to build in Clang for FVP because of the ``RECLAIM_INIT_CODE`` option:
```
aarch64-linux-gnu-ld.bfd: Init code ends past the end of the stacks
```
I guess that we have to remove the ``__init`` attribute from some functions...

**Comment by soby-mathew** [comment]
Yes, this is a FVP problem not related to the patch, which we can handle separately. @lmayencourt, is looking into this.

@antonio-nino-diaz-arm , are you happy with the changes otherwise ?

**Comment by antonio-nino-diaz-arm** [comment]
Fine by me.

**Pull request merged**